### PR TITLE
CORGI-811 stream multi variants

### PR DESCRIPTION
**Pull request by jasinner**
This addressed CORGI-811 by only attaching variants from errata_info to active streams.

It also removes some logic from brew_tag parsing which skipped some streams. This is no longer required as we don't associate variants with these streams any more.

@RedHatProductSecurity/corgi-devs please review.

### CODE REVIEW

File: corgi/tasks/prod_defs.py [hunk 1]
```diff
@@ -323,22 +323,7 @@ def _parse_cpes_from_brew_tags(
     version_name: str,
 ) -> list[str]:
     """Match streams using brew_tags to Errata Tool Product Versions and their Variants"""
```
Comment (juspence) [inline corgi/tasks/prod_defs.py]: Maybe a small update to the comment is helpful:
```suggestion
    """Match streams using brew_tags to Errata Tool Product Versions and their Variants
       Return a list of CPEs for the matched Variants"""
```

File: corgi/tasks/prod_defs.py [hunk 2]
```diff
@@ -323,22 +323,7 @@ def _parse_cpes_from_brew_tags(
...
-    # quay-3 streams brew_tags all share a single variant, and we have a one-to-many
-    # cardinality between streams and variants. quay-3 streams should probably be consolidated
-    # into a single stream in prod_defs. See also PROJQUAY-5312.
-    # The rhn_satellite_6 streams have brew tags, but those brew tags are associated
-    # with the RHEL-7-SATELLITE-6.10 ET Product Version. We skip them
-    # here to ensure the rhn_satelite_6.10 variants only get linked with that stream
-    # I haven't filed a product bug to get them fixed since 6.7 - 6.9 are no longer
-    # active. See CORGI-546 for more details.
-    # Also skip brew_tag matching for dts ps_module which share Variants/Brew Tags with rhscl
-    # See CORGI-726
```
Comment (juspence) [inline corgi/tasks/prod_defs.py:335]: For Satellite, the older streams are no longer active so I think this change doesn't matter. But if we stop skipping the DTS stream, I think we'll end up reporting whatever CPEs we find here in the manifest for that stream. Will that give the correct results, or do we need to keep skipping DTS here?

File: corgi/tasks/prod_defs.py [hunk 3]
```diff
@@ -399,7 +388,7 @@ def parse_errata_info(
                         },
                     },
                 )
-                ProductNode.objects.get_or_create(
+                ProductNode.objects.update_or_create(
```
Comment (juspence) [inline corgi/tasks/prod_defs.py]: Maybe a short comment is helpful to remind us in the future:
```suggestion
                # If multiple active streams share the same errata_info,
                # use update to relink the node to a different parent product_stream_node
                # the last matching stream always wins / gets the link
                ProductNode.objects.update_or_create(
```
This feels a little risky - moving the nodes around caused some bugs before, and I still don't really understand why.

Can we check to see how many variants / errata_info are shared by different streams? I hope that the case you found is the only example. Or at least it's always one active z-stream / one inactive base stream that share variants.

But maybe there's some weird edge case where multiple active streams share the same variant. I think in that case, Variant-type ProductComponentRelations will get processed and components may get linked to the wrong variant, or to both variants if we saved their taxonomies before this change.

Finally, if we linked the variant to stream A before this change, it's possible this new code will relink the variant to stream B instead. So it may trigger some unwanted updates the next time the task runs.

File: corgi/tasks/prod_defs.py [hunk 4]
```diff
@@ -369,6 +354,10 @@ def parse_errata_info(
     """Parse and create ProductVariants from errata_info in product-definitions.json"""
     et_product_versions_set = set(product_stream.et_product_versions)
     for et_product in errata_info:
+        # This is a workaround for CORGI-811 which attaches variants from errata_info only to active
+        # stream
+        if not product_stream.active:
+            continue
```
Comment (juspence) [inline corgi/tasks/prod_defs.py]: I think this should be either:

1. an early return outside the for-loop, since the check will always be the same inside the loop, and we won't do anything inside it if we skip processing this early.

2. OR a continue atter line 366 / `et_product_versions_set.add(et_pv_name)`, so we continue to update et_product_versions even for inactive streams, but without creating and linking any variants.

Unrelated to your changes, is there a small bug on line 355? We always append to this stream's list of et_product_versions. So I think Corgi will still report any old PVs even if they were removed (when data in prod-defs changes).

We could start processing with an empty set, and then add any current PVs to this stream only if they're still present in prod-defs (as of the day this task runs).

File: tests/test_products.py [hunk 5]
```diff
@@ -534,3 +484,77 @@ def test_brew_tag_matching(variant_name, cpe, brew_tag, stream_name, requests_mo
...
+        requests_mock.get(f"{settings.PRODSEC_DASHBOARD_URL}/product-definitions", text=text)
+
+    update_products()
+
+    ga_stream.refresh_from_db()
+    assert ga_stream.productvariants.get_queryset().count() == 0
+    z_stream.refresh_from_db()
+    assert z_stream.productvariants.get_queryset().count() == 1
+    assert not ProductComponentRelation.objects.filter(software_build=sb).exists()
+    assert mock_remove.called_with(
```
Comment (juspence) [inline tests/test_products.py]: Maybe below is better, or maybe it's expected that the mock is called multiple times?
```suggestion
    assert mock_remove.called_once_with(
```

File: tests/test_products.py [hunk 6]
```diff
@@ -534,3 +484,77 @@ def test_brew_tag_matching(variant_name, cpe, brew_tag, stream_name, requests_mo
...
+            z_stream.pk,
+        )
+    )
+
+
+@pytest.mark.django_db
+def test_multi_variant_active_streams(requests_mock):
+    """Tests that multiple runs of update_products with multiple active streams sharing errata_info
+    does not cause the variants to move back and forth between the streams"""
+    with open("tests/data/prod_defs/proddefs-update-multi-stream-variant-active.json") as prod_defs:
```
Comment (juspence) [inline tests/test_products.py:565]: Thanks for adding this! Should we also add a similar test, where the first call to update_products uses a different test file than the second?

This would simulate a prod-defs change, where some stream starts using the same errata_info as another stream (when it wasn't before).

File: tests/test_products.py [hunk 7]
```diff
@@ -534,3 +484,77 @@ def test_brew_tag_matching(variant_name, cpe, brew_tag, stream_name, requests_mo
...
+
+    update_products()
+
+    ga_stream = ProductStream.objects.get(name="ga_stream")
+    assert ga_stream.productvariants.get_queryset().count() == 0
+    z_stream = ProductStream.objects.get(name="z_stream")
+    assert z_stream.productvariants.get_queryset().count() == 1
+
+    with open("tests/data/prod_defs/proddefs-update-multi-stream-variant-active.json") as prod_defs:
+        text = prod_defs.read()
```
Comment (juspence) [inline tests/test_products.py]: I think you can just reuse the values stored in the text variable, you should't need to open the same file and read the same data again. We just need to add the second request to the mock.

File: tests/test_products.py [hunk 8]
```diff
@@ -436,56 +436,6 @@ def test_find_by_cpe():
     assert cpe in results
 
 
-@pytest.mark.django_db
-def test_find_by_cpe_substring():
```
Comment (juspence) [inline tests/test_products.py]: I get why we don't need the test for skipping buggy linking, but should this test be kept? I don't know where or how we use the _find_by_cpe method it's testing.

**Review by juspence — APPROVED** [review]
I have some concerns about rare edge-cases (as always), but maybe they never happen and it's fine as-is. Overall LGTM.

File: corgi/tasks/prod_defs.py (hunk 4 above)
Comment (jasinner) [reply corgi/tasks/prod_defs.py]: Good idea I missed the fact that we are adding to the et_product_versions_set on line 366, thanks.

Good idea bout change 355 to an empty set.

File: corgi/tasks/prod_defs.py (hunk 2 above)
Comment (jasinner) [reply corgi/tasks/prod_defs.py:335]: The cpes_from_brew_tags property now has the correct CPEs for dts streams.

File: corgi/tasks/prod_defs.py (hunk 3 above)
Comment (jasinner) [reply corgi/tasks/prod_defs.py]: I wrote a little skip to check prod_defs data and found that the only variant which shares streams is the one I previously mentioned `8Base-CertSys-10.4`. I have a test for multiple active streams, and verified the variant didn't move when multiple update_product tasks were run. The only issue that could happen is if the order of the streams was changed in product_definitions. But I think the likelyhood of that happening is so low we should probably ignore it for now.

```
streams_by_variant = defaultdict(list)
for product in ProdDefs.load_product_definitions():
    for version in product["product_versions"]:
        for stream in version["product_streams"]:
            if "errata_info" in stream:
                for errata_product in stream["errata_info"]:
                    for errata_version in errata_product["product_versions"]:
                        for errata_variant in errata_version["variants"]:
                            streams_by_variant[errata_variant].append(stream["id"])
```

File: tests/test_products.py (hunk 8 above)
Comment (jasinner) [reply tests/test_products.py]: Good catch. Sorry I removed that by accident. Adding it back.

File: tests/test_products.py (hunk 5 above)
Comment (jasinner) [reply tests/test_products.py]: fixed.

File: tests/test_products.py (hunk 6 above)
Comment (jasinner) [reply tests/test_products.py:565]: That's what the other test `test_multi_variant_streams` does.

File: tests/test_products.py (hunk 7 above)
Comment (jasinner) [reply tests/test_products.py]: Good idea, fixed.

File: corgi/tasks/prod_defs.py (hunk 1 above)
Comment (jasinner) [reply corgi/tasks/prod_defs.py]: updated

File: tests/test_products.py (hunk 6 above)
Comment (juspence) [reply tests/test_products.py:565]: That test is using one inactive stream and one active stream. I was thinking here we should add a test for two different streams which are both active. But you confirmed above this doesn't happen today, so I think it's fine.

If it does start happening, I think we'll see churn with Variants constantly moving back and forth between streams. Fingers crossed this won't be a problem...
